distributed/scheduler.py

```diff
@@ -1096,6 +1106,15 @@ def _to_dict_no_nest(self, *, exclude: Container[str] = ()) -> dict[str, Any]:
         """
         return recursive_to_dict(self, exclude=exclude, members=True)
 
+    @property
+    def rootish(self):
+        return (
+            not self._has_restrictions
+            and len(self) >= 5
+            and len(self.dependencies) < 5
+            and sum(map(len, self.dependencies)) < 5
```

Comment:
```suggestion
            and (ls := len(self)) >= 5
            and len(self.dependencies) < 5
            and (ndeps := sum(map(len, self.dependencies))) < 5
            and ls / ndeps > 2
```
Maybe we're trying to get at the idea of "fanning out"? As in there are significantly (say 2x) more tasks in the group than there are inputs?

idk if this is valuable, just a thought. I think it would also eliminate the need for the magic `len(self) >= 5`.

Reply:
I am a bit hesitant here because we don't have any examples (i.e. tests) that indicate the one or the other would be favorable. At least in the future we should motivate changes to this by new/modified tests.

Reply:
If we went for your logic, why even include `len(self) >= 5`? Why not just

```python
(
            len(self.dependencies) < 5
            and (ndeps := sum(map(len, self.dependencies))) < 5
            # Fan-out
            and (len(self) / ndeps > 2 if ndeps else True)
        )
```

(FWIW this would also include the case of no deps naturally

Reply:
What I don't like about this suggestion is that there is again another magic number `2` in there. Why would a fan-out only be a fan-out if there is a factor of 2? For entirely symmetric cases I can see how we can motivate 2 but not for asymmetric ones, e.g.

```mermaid
graph BT;
  A1-->B1;
  A1-->B2;
  A2-->B3;
  A2-->B4;
  A3-->B5;
```

Is this meaningfully different to the case where A3 splits into two tasks as well? Anyhow, I believe arguing about these small graphs is not truly relevant. As soon as we hit any real world graphs these magic numbers are getting more and more meaningless

Reply:
> If we went for your logic, why even include `len(self) >= 5`?

Sorry, I meant to remove that check, just forgot to delete it in the diff. Agreed that it's unnecessary.

> For entirely symmetric cases I can see how we can motivate 2 but not for asymmetric ones, e.g.

This gets at a larger problem, which is the use of TaskGroups at all for these types of scheduling choices. We're assuming that TaskGroups indicate something structural about the graph, but they don't necessarily. I think we have to accept that we're assuming TaskGroups are:
* symmetrical
* correspond 1:1 with structural layers in the graph

If we're going to use TaskGroups for these sorts of heuristics, we have to be okay taking on those assumptions. I'd rather focus on graphs that meet these assumptions, but still would schedule poorly.

Reply:
> I'm not even sure if "fanning out" is something we want to capture

I'm not sure either. Though fanning out is what the "ish" part of root-ish task logic was trying to capture. The idea was that when a tiny number of tasks "fan out" to a much wider number of tasks, if you don't copy the input tasks to most workers, you'll end up highly imbalanced, with most workers idle and a couple overwhelmed.

Mathematically though, I'm realizing that so long as we have `sum(map(len, self.dependencies)) < 5`, whether we call it fan-out, or a lower bound on `len(self)`, they're effectively equivalent. They're both expressing "there are very few tasks that go into the task group, and the task group is decently larger than the number of inputs". It's just arithmetic to turn one into the other.

> What I don't like about this suggestion is that there is again another magic number `2` in there

Yup, I don't like that either. I don't think we'll be able to come up with a _static_ definition for "should be queued" that doesn't include a magic number though, at least without significantly refactoring how the queued state works.